test/jdk/java/util/zip/ZipFile/ZipSourceCache.java

```diff
@@ -0,0 +1,162 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8317678
+ * @modules java.base/java.util.zip:open
+ * @summary Fix up hashCode() for ZipFile.Source.Key
+ * @run junit/othervm ZipSourceCache
+ */
+
+import java.io.*;
+import java.lang.reflect.*;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.*;
+import java.util.zip.*;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.AfterAll;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class ZipSourceCache {
+
+    private static final String ZIPFILE_NAME =
+            System.currentTimeMillis() + "-bug8317678.zip";
+    private static final String ZIPENTRY_NAME = "random.txt";
+    private static final String INVALID_LOC_EXCEPTION =
+            "ZipFile invalid LOC header (bad signature)";
+    private static final boolean DEBUG = false;
+
+    private static File relativeFile = new File(ZIPFILE_NAME);
+    private static File absoluteFile = new File(ZIPFILE_NAME).getAbsoluteFile();
+    private static boolean hasfileKeySupport;
+
+    @BeforeAll
+    public static void setup() throws Exception {
+        createZipFile("test1");
+        var attrs = Files.readAttributes(relativeFile.toPath(), BasicFileAttributes.class);
+        hasfileKeySupport = (attrs.fileKey() != null);
+    }
+
+    @AfterAll
+    public static void cleanup() throws IOException {
+        Files.deleteIfExists(Path.of(ZIPFILE_NAME));
+    }
+
+    /*
+     * Monitor the internal "files" HashMap to ensure that we only
+     * create one <Key, Source> mapping per unique zip file.
+     *
+     * This test also ensures that a new <Key, Source> mapping is created
+     * if an update to an existing zip file is detected.
+     */
+    @Test
+    public void testKeySourceMapping() throws Exception {
+        ZipFile absoluteZipFile;
+        HashMap internalMap;
+        int numSources;
+        try (ZipFile zipFile = new ZipFile(ZIPFILE_NAME)) {
```

Comment:
Do you think changing this to:
```java
new ZipFile(relativeFile)
```
would be better and more clear about the intent? A few lines below, we use `new ZipFile(absoluteFile)` which I think is good because it makes it clear what is being done.